cpp/src/arrow/array/array_nested.cc

```
@@ -683,7 +683,7 @@ void SparseUnionArray::SetData(std::shared_ptr<ArrayData> data) {
   ARROW_CHECK_EQ(data_->buffers.size(), 2);
 
   // No validity bitmap
-  ARROW_CHECK_EQ(data_->buffers[0], nullptr);
+  data_->buffers[0] = nullptr;
```

Hmm, if something is passing a validity bitmap, I think the caller should be fixed. Else, presumably the kernel is generating a validity bitmap that it expects to be used, but which gets silently ignored?

---

That may be so. I was thinking, constructing UnionArray or otherwise UnionArray::SetData must have the authority to ensure validity bitmap is null given it's in the specification that they don't have validity bitmaps?

---

b/c as noted, the kernel `struct_field` and calling `flatten` both cause the crash.

---

I would say that's a bug in the kernel, and that the assertion here is precisely to catch that kind of error.

---

(it should combine the parent null bitmap with the bitmap of each of its childs ..?)

---

…yeah, that sounds right (that sounds like a pain!)

---

Agreed that it's a bug that needs to be fixed.